### ISSUE TITLE
[ENH] Share external C libraries across multiple Cython modules

### ISSUE THREAD

**Issue by Keithcat1** [cython/cython#4682]
**Is your feature request related to a problem? Please describe.**
When wrapping a C library for Python, there are two ways to ship it. If you statically link it, you have to put all of your bindings in a single extension module which is uncomfortable, forces you to either put everything in one file or use the include statement or risk statically linking the library into each extension module that  uses it, and prevents you conveniently separating things out into their own modules.
On the other hand, you can dynamically link the library and ship the DLL. I'm currently using Scikit-build though, which while it makes it much, much easier to build complicated packages easily cross-platform, appears to have no way to make the DLLs findable on the path on Linux. On Windows at least works, though.

**Describe the solution you'd like**
When a PXD file uses:
```cython
cdef extern from "myheader.h": pass
```
in order to import declarations from a C header, and that PXD module is not alone (I.E. is associated to a .PYX module) I want a way to specify that all declarations imported from the header in the cdef extern from block should be made cimportable by other Cython modules. This would also mean that C functions would have to be exported via __pyx_capi__. This would allow you to create an extension module which statically links the C library you're wrapping which can then easily be distributed.

**Describe alternatives you've considered**

**Additional context**

**Comment by scoder** [comment]
I think a better alternative is to link multiple extension modules (e.g. a whole package) into one.

https://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html#integrating-multiple-modules

So, you can either create a single shared library and have all your Cython modules use it, or create a single extension module and link your C library statically into that.

I'd rather not add a third, more complex option.

Also see #2356 and #2011.

**Comment by scoder** [comment]
> https://docs.cython.org/en/latest/src/userguide/source_files_and_compilation.html#integrating-multiple-modules

Admittedly, that section doesn't explain how to do it from inside of extension modules. There's more to add to the documentation here, following

https://docs.python.org/3/c-api/module.html#initializing-c-modules

Let's reopen this and consider it a documentation issue.

**Comment by da-woods** [comment]
Personally I'd probably write some very small wrapper `cdef` functions (which could then be shared with the `__pyx_capi__` mechanism). I know it's a bit of duplication, but it's easy easy for the user to do (so I agree with the "wontfix" label - there's an easy workaround so we don't need to complicate Cython much further)

**Comment by da-woods** [comment]
The big reason not to do this automatically: sometimes people declare c functions that don't really exist (for example, may only be present in a certain library version, or may actually be a macro, or similar). If we automatically added extern functions to `__pyx_api__` this cases would break.

**Comment by Keithcat1** [comment]
Yeah. I realized that PyImport_AppendInittab should be called before PyInitialize, so it's probably not suitable for this.
